xarray/core/dataset.py

```diff
@@ -1491,7 +1501,19 @@ def __setitem__(self, key: Hashable, value) -> None:
                 "cannot yet use a dictionary as a key to set Dataset values"
             )
 
-        self.update({key: value})
+        if isinstance(key, list):
+            if len(key) != len(value):
+                raise ValueError(
+                    f"Different lengths of variables to be set "
+                    f"({len(key)}) and data used as input for "
+                    f"setting ({len(value)})"
+                )
+            if isinstance(value, Dataset):
+                self.update(dict(zip(key, value.data_vars.values())))
+            else:
+                self.update(dict(zip(key, value)))
```

Comment:
I think the case `ds[["a"]] = xr.DataArray([1, 2]))` is unlikely but dangerous here. Maybe:

```suggestion
                if isinstance(value, DataArray):
                    raise ValueError("Cannot assign single DataArray to multiple keys")
                self.update(dict(zip(key, value)))
```

Reply:
Nice, good spot @mathause

Reply:
I'm guessing you mean `ds[["a","b"]] = xr.DataArray([1, 2]))` since the error message is about multiple keys :)

I have added some logic that makes `ds[["a"]] = xr.DataArray([1, 2]))` possible (unless I'm missing some obvious problem with this).